File: OlderCodes/Ccodes/diff.c
/* 
************************************************************************
*  diff.c:  Differentiation with forward, central and  extrapolated    *
*            difference methods                                        *
*                                                                      *
*  From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
   by RH Landau, MJ Paez, and CC BORDEIANU 
   Copyright Princeton University Press, Princeton, 2008.
   Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
   MJ Paez, Univ Antioquia, 2008; & CC BORDEIANU, Univ Bucharest, 2008
   Support by National Science Foundation                              
*                                                                      *
*  comment: results saved as x y1 y2 y3                                *
************************************************************************
*/
#include <stdio.h>
#include <math.h>

#define h 1e-5                          /* stepsize for all methods */
#define xmax 7                        /* range for calculation */
#define xmin 0
#define xstep 0.01                  /* stepsize in x  */

main()
{ 
   double dc, result, x;
   double f(double);      /* function we differentiate */ 
          /* see end of code */             
   FILE *output;      /* save data in diff.dat */
   output = fopen("diff.dat","w");
   
   for (x=xmin; x<=xmax; x+=xstep)
   {
      fprintf(output,"%f\t", x);
 
      result=(f(x+h)-f(x))/h;                 /* forward difference */
      fprintf(output, "%.10f\t", result);
   
      result=(f(x+h/2)-f(x-h/2))/h;           /* central difference */
      fprintf(output, "%.10f\t", result);
   
      result=(8*(f(x+h/4.0)-f(x-h/4.0))-(f(x+h/2)-f(x-h/2.0)))/(3.0*h);
      fprintf(output, "%.10f\n", result);     /* extrapolated diff */
   }
   printf("data stored in diff.dat\n");
   fclose(output);
}
/*---------------------end of main program----------------------------*/

/* the function we want to differentiate */
double f(double x)    
{
   return(cos(x));
}

File: OlderCodes/JavaCodes/Diff.java
/*
*****************************************************************************
* Diff.java: Differentiation with forward, central and  extrapolated        *
*            difference methods                                             *
*  From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
   by RH Landau, MJ Paez, and CC Bordeianu 
   Copyright Princeton University Press, Princeton, 2007.
   Electronic Materials copyright: R Landau, Oregon State Univ, 2007;
   MJ Paez, Univ Antioquia, 2007; & CC Bordeianu, Univ Bucharest, 2007
   Support by National Science Foundation                              
* 
* comment: results saved as x y1 y2 y3                                      *                                                                    
***************************************************************************** 
*/


import java.io.*;


public class Diff
 {     
public static double h= 1e-5;                       /* stepsize for all methods */
public static int xmax =7 ;                         /* range for calculation */
public static int xmin =0;
public static double xstep =0.01 ;                  /* stepsize in x  */

public static void main(String[] argv) throws IOException,
FileNotFoundException
{ 
   double dc, x;
   double result[]= new double[3];
                                        
   /* save data in diff.dat */
   
   PrintWriter w = 
        new PrintWriter(new FileOutputStream("diff.dat"), true);
   for (x=xmin; x<=xmax; x+=xstep)
   {
      result[0]= (f(x+h)-f(x))/h;              /* forward difference */
      result[1]=(f(x+h/2.)-f(x-h/2.))/h;/* central difference */
      result[2]=(8.*(f(x+h/4.)-f(x-h/4.))-(f(x+h/2.)-f(x-h/2.)))/(3.*h);
     /* extrapolated diff */
      w.println(""+ x+"  "+result[0]+"  "+result[1]+"  "+result[2]+"  ");
 
   }
   System.out.println("data stored in diff.dat");
   
}
/*---------------------end of main program----------------------------*/

/* the function we want to differentiate */
public static double f(double x)        
{
   return(Math.cos(x));
}
}
